src/main/java/edu/hm/hafner/analysis/ModuleDetector.java

```diff
@@ -181,6 +202,56 @@ private String parseBuildXml(final String buildXml) {
         return StringUtils.EMPTY;
     }
 
+    /**
+     * Returns the project name estimated from the build.gradle file path.
+     *
+     * @param buildScript
+     *         Gradle build.gradle file path
+     *
+     * @return the project name or an empty string if the name could not be resolved
+     */
+    private String parseGradle(final String buildScript) {
+        Path path = Paths.get(buildScript).normalize();
+        if (path.getNameCount() >= 2) {
+            Path parent = path.getParent();
+            return parent.getFileName().toString();
+        }
+        return StringUtils.EMPTY;
+    }
+
+    /**
+     * Returns the root project name from the settings.gradle file.
+     *
+     * @param settingsFile
+     *         Gradle settings.gradle file path
+     *
+     * @return the root project override, or an empty string if the name could not be resolved
+     */
+    private String parseGradleSettings(final String settingsFile) {
+        String name = null;
+
+        try (InputStream input = factory.open(settingsFile);
+             Scanner scan = new Scanner(input, "UTF-8")
+        ) {
+            Pattern namePattern = Pattern.compile("\\s*rootProject\\.(name\\s*=|setName[(]?)\\s*['\"](.*?)['\"][)]?");
```

Comment:
This pattern is a simple way to handle both `'` and `"` strings, property-setter syntax (`name = "..."`), and setter syntax (`setName("...")`), as well as the groovy parens-optional syntax of the setter (`setName "..."`).

The pattern is not perfect:  i.e., it does not match open/close quotes, so technically `name = '.."` would pass this pattern, even though it's invalid in groovy.  Similarly for the setter parens, where it does not require either no parens or both parens, and instead considers each of open and close to be optional, even though that would not be valid Groovy.

Reply:
It make sense to extract the pattern to a constant: there you can add your valuable explanation as a comment so it is not lost.

Reply:
It make sense to extract the pattern to a constant: there you can add your valuable explanation as a comment so it is not lost.

Reply:
Can you please additionally check if you can improve the regexp to make it save for ReDOS attacks (see https://ci.jenkins.io/job/Plugins/job/analysis-model/job/PR-362/6/spotbugs/new/source.a0b988b1-e861-4191-89b9-9f5100768de2/#236)

Reply:
I believe the ReDOS was addressed by this change:

```diff
"\\s*rootProject\\.(name\\s*=|setName[(]?)\\s*['\"](.*?)['\"][)]?"
"^\\s*rootProject\\.(name\\s*=|setName\\(?)\\s*['\"]([^'\"]*)['\"]\\)?"
```
1. Start with `^`, to avoid the issues with `\s*`
2. replace `[(]?` with `\\(?` and `[)]?` to `\\)?` -- I'm not sure if this was really an issue
3. Replace `(.*?)` with `[^'"]*` -- gives a more clear expectation for when the pattern will stop. Gradle refuses to accept: `[/, \, :, <, >, ", ?, *, |]`, but it is possible to add `'` to the project name.  If we're concerned about being 100% strict, we're probably better off reading the Gradle model after configure.

I believe (3) and (1) were the main reasons for the ReDOS warnings